### ISSUE TITLE
error handling cleanups

### ISSUE THREAD

**Issue by Blub** [lxc/lxc#2471]
We have a bunch of functions throughout with inconsistent error return values, mixing `-1` with and without a valid `errno` with returning negative errno codes. Eg. some confile.c functions can return `-EINVAL` as well as just `-1`.
We should clean these up to follow the coding style[1] document.
In other words: in our own functions, if not in conflict with a successful return value, always return a negative error code (iow, if the error was triggered by something setting `errno`, simply `return -errno;`), otherwise actually set `errno` if needed. (Eg. when a non-NULL pointer is to be returned on success, return `NULL` after setting `errno`).

[1] https://github.com/lxc/lxc/blob/master/CODING_STYLE.md#return-error-codes

**Comment by brauner** [comment]
@2xsec, this might be something you might be interested in doing. Probably a good idea to either do a commit for each function or for each file.

**Comment by 2xsec** [comment]
@brauner, Of course. I'll check it.

**Comment by brauner** [comment]
@2xsec, no pressure just in case you're interested. You might also consider jumping into our irc channel #lxc-dev on Freenode. But of course, that's completely optional too. :)

**Comment by 2xsec** [comment]
@brauner This is no pressure for me.:) Please recommend a good tool for irc.:)

**Comment by brauner** [comment]
Hm, we usually either use `irssi` or `weechat`. But there are graphical or browser based irc clients as well. It depends on whether you want to be connected constantly or only on demand. Most of us have a persistent irc session running in tmux on a remote server so that we can simply attach to it every day and see what we missed and also have logs of all interactions.
Personally, I use ZNC as irc bouncer but that is probably not needed.

**Comment by Blub** [comment]
Updated my intial comment a bit in response to #2551. I think this part might benefit from a few examples in the coding style doc, or some more situation distinctions.

**Comment by brauner** [comment]
@Blub, cool. From what I gathered from our recent irc discussion you'd prefer to set `errno` instead of returning `-<errno-value>` from functions, right. I'll let you review branches such as https://github.com/lxc/lxc/pull/2551 since you have a better grasp of what you're aiming for. :)

**Comment by Blub** [comment]
I don't remember saying that ;-). Setting `errno` was for functions where the return type conflicts with the error codes (eg pointers.). Weren't you entertaining the idea of using `PTR_ERR()`? That one I'm not so fond of ;-).

**Comment by brauner** [comment]
No, `PTR_ERR()` and friends should stay restricted to our kernel work. ;) We agreed on that on irc a while back. `errno` is what we have for the userspace part of our work.

**Comment by 2xsec** [comment]
@brauner @Blub I'm sorry. I didn't finish this issue because there are so many functions return int or errno values. So, I think we need to define well-defined return type. If you have any good idea, I'll try to work for this. Thanks. :)

**Comment by brauner** [comment]
I'll let @Blub spearhead this work and its direction. I'm just going to give snide comments from time to time. ;)

**Comment by 2xsec** [comment]
@brauner Thanks. :)

**Comment by tcharding** [comment]
I'll close #2551 then and let you boys at it.

**Comment by stgraber** [comment]
@brauner do we still need this issue or did you accidentally clean that all up by now?
